src/audio_input/audio.c

```diff
@@ -0,0 +1,186 @@
+#define AUDIO_DEVICE_ADD_WAIT_TIME_MS 100
+
+static void audioCallback(void *userdata, uint8_t *stream, int len) {
+   RSAudio *audio = (RSAudio *)userdata;
+   rsCircleStaticAdd(&audio->data, stream, (size_t)len);
+}
+
+static char *getDeviceName(const char *devname) {
+   const char *monitor = "Monitor of ";
+   if (!devname || !strcmp(devname, "auto")) {
+      return NULL;
+   }
+   if (!strcmp(devname, "default")) {
+      const char *sdlname = SDL_GetAudioDeviceName(0, false);
+      if (!sdlname) {
+         return NULL;
+      }
+      char *newname = rsMemoryCreate(strlen(sdlname) + strlen(monitor) + 1);
+      strcpy(newname, monitor);
+      strcat(newname, sdlname);
+      return newname;
+   }
+   return rsStringClone(devname);
+}
+
+static void probeDevices(RSAudio *audio) {
+   int max = SDL_GetNumAudioDevices(true);
+   rsLog("Available devices:");
+   for (int i = 0; i < max; ++i) {
+      rsLog(" - %s", SDL_GetAudioDeviceName(i, true));
+   }
+   audio->deviceNum = max;
+   SDL_GetNumAudioDevices(false);
+}
+
+static bool openDevice(RSAudio *audio, const char *devname) {
+   SDL_zero(audio->ospec);
+   audio->deviceId = SDL_OpenAudioDevice(devname, true, &audio->ispec, &audio->ospec, 0);
+   if (!audio->deviceId) {
+      rsLog("SDL2 couldn't open audio device %s : %s", SDL_GetError(), devname);
+      return false;
+   }
+   rsLog("Opened audio device %s", devname);
+   return true;
+}
+
+void rsAudioCreate(RSAudio *audio, const RSConfig *config) {
+   if (SDL_Init(SDL_INIT_AUDIO | SDL_INIT_EVENTS) < 0) {
+      rsError("SDL2 could not initialize! SDL Error: %s", SDL_GetError());
+   }
+   SDL_EventState(SDL_AUDIODEVICEADDED, SDL_ENABLE);
+   SDL_EventState(SDL_AUDIODEVICEREMOVED, SDL_ENABLE);
+   if (pthread_spin_init(&audio->sampleGetLock, PTHREAD_PROCESS_PRIVATE)) {
+      rsError("pthread_spin_init failed");
+   }
+
+   SDL_zero(audio->ispec);
+   audio->ispec.freq = (int)config->audioSamplerate;
+   audio->ispec.format = AUDIO_S16LSB;
+   audio->ispec.channels = (Uint8)config->audioChannels;
+   audio->ispec.samples = SAMPLES_PER_CALLBACK; // must be power of 2
+   audio->ispec.callback = audioCallback;
+   audio->ispec.userdata = audio;
+
+   probeDevices(audio);
+   audio->deviceName = getDeviceName(config->audioDeviceName);
+   if (!openDevice(audio, audio->deviceName)) {
+      rsError("Exiting");
+   }
+
+   int size1s = audio->ospec.channels * audio->ospec.freq * (int)sizeof(uint16_t);
+   int sizeTotal = size1s * config->duration;
+   int numOfCallbacks =
+       (sizeTotal + (SAMPLES_PER_CALLBACK - 1)) / SAMPLES_PER_CALLBACK; // round up div
+   sizeTotal = numOfCallbacks * SAMPLES_PER_CALLBACK;
+   audio->sizeBatch = size1s / config->framerate;
+   rsCircleStaticCreate(&audio->data, (size_t)sizeTotal);
+   audio->deviceAddTime = 0;
+   audio->deviceRemoveTime = 0;
+   SDL_PauseAudioDevice(audio->deviceId, 0);
+}
+
+static void rsAudioReconnect(RSAudio *audio, const char *devname) {
```

Comment:
Remove `rs` and move it up.

Reply:
done

> Did some more testing and the `Monitor of` trick doesn't seem to work since my first output device isn't my default and [sadly that does not seem to be possible](https://discourse.libsdl.org/t/get-audio-device-name-from-sdl-audiodeviceid/22667/2). However I'm happy explicitly setting my device if there is hotplugging available.
> 
> The hotplugging itself works very well, although the audio is very out of sync.

Is it out of sync all the time or only in the first 'duration' seconds ? It should sync up once the buffer gets filled up